src/pgas.jl

```diff
@@ -0,0 +1,201 @@
+Must return `nothing` if it consumed all the data
+"""
+function observation end
+
+"""
+    inc_step!(model::AbstractStateSpaceModel, n::Int=1)
+
+Increase the internal model counter by `n`. Used to keep track of the current timestep when replaying
+"""
+inc_step!(m::AbstractStateSpaceModel, n::Int=1) = m.T += n
+
+"""
+    reset!(m::AbstractStateSpaceModel)
+
+Set model counter to `1` to restart model evaluation
+"""
+reset!(m::AbstractStateSpaceModel) = m.T = 1
+
+"""
+    previous_state(model::AbstractStateSpaceModel)
+
+Return `Xₜ₋₁` or `nothing` from `model`
+"""
+function previous_state(model::AbstractStateSpaceModel)
+    return model.T < 2 ? nothing : model.X[model.T - 1]
+end
+
+"""
+    current_step(model::AbstractStateSpaceModel)
+
+Return current model step
+"""
+current_step(model::AbstractStateSpaceModel) = model.T
+
+"""
+    transition_logweight(model::AbstractStateSpaceModel, x)
+
+Get the log weight of the transition from previous state of `model` to `x`
+"""
+function transition_logweight(particle::SSMTrace, x)
+    return Distributions.logpdf(transition(particle.f, previous_state(particle.f)), x)
+end
+
+"""
+    get_ancestor_logweights(pc::ParticleContainer{F,R}, x) where {F<:SSMTrace,R}
+
+Get the ancestor log weights for each particle in `pc`
+"""
+function get_ancestor_logweights(pc::ParticleContainer{F,R}, x) where {F<:SSMTrace,R}
+    particles = collect(pc.vals)
+    nparticles = length(particles)
+    weights = pc.logWs
+
+    logweights = map(1:nparticles) do i
+        transition_logweight(particles[i], x) + weights[i]
+    end
+
+    return logweights
+end
+
+""" 
+    advance!(particle::SSMTrace, isref::Bool=false)
+
+Return the log-probability of the transition nothing if done
+"""
+function advance!(particle::SSMTrace, isref::Bool=false)
+    isref ? load_state!(particle.rng) : save_state!(particle.rng)
+
+    model = particle.f
+    if !isref
+        if model.T == 1
+            new_state = rand(particle.rng, initialization(model)) # Generate initial state, maybe fallback to 0 is initialization is not defined
+        else
+            current_state = model.X[model.T - 1]
+            new_state = rand(particle.rng, transition(model, current_state))
+        end
+    else
+        new_state = model.X[model.T <= length(model.X) ? model.T : end] # Hacky, reached beyond last time step
+    end
+
+    score = observation(model, new_state)
+
+    if score !== nothing # Accept transition
+        !isref && push!(model.X, new_state)
+        inc_step!(model)
+        inc_counter!(particle.rng) # Increase rng counter, don't really need it but keep things consistent
+    else
+        inc_step!(model, -1)
+    end
+
+    return score
+end
+
+function truncate!(model::AbstractStateSpaceModel)
+    step = model.T <= length(model.X) ? model.T - 1 : length(model.X)
+    return model.X = model.X[1:step]
+end
+
+function fork(trace::SSMTrace, isref::Bool)
+    model = deepcopy(trace.f)
```

Comment:
I assume this might become memory intensive?

Reply:
Yes, this is the main issue with this implementation, we keep copying state vectors around which will probably be quite heavy for long time series